did-ion/src/sidetree.rs

```diff
@@ -0,0 +1,1926 @@
+    // List of error codes: https://github.com/decentralized-identity/sidetree/blob/v1.0.0/lib/core/versions/1.0/ErrorCode.ts
+    pub code: String,
+    pub message: Option<String>,
+}
+
+impl fmt::Display for SidetreeError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Sidetree error {}", self.code)?;
+        if let Some(ref message) = self.message {
+            write!(f, ": {}", message)?;
+        }
+        Ok(())
+    }
+}
+
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+impl<S: Sidetree + Send + Sync> DIDMethod for SidetreeClient<S> {
+    fn name(&self) -> &'static str {
+        S::METHOD
+    }
+
+    fn to_resolver(&self) -> &dyn DIDResolver {
+        match self.resolver {
+            Some(ref res) => res,
+            None => &NoOpResolver,
+        }
+    }
+
+    fn create(&self, create: DIDCreate) -> Result<DIDMethodTransaction> {
+        let DIDCreate {
+            recovery_key,
+            update_key,
+            verification_key,
+            options,
+        } = create;
+        ensure!(is_empty(&options), "Create options not supported");
+        let (update_pk, recovery_pk, patches) =
+            new_did_state::<S>(update_key, recovery_key, verification_key)
+                .context("Prepare keys for DID creation")?;
+        let operation = S::create_existing(&update_pk, &recovery_pk, patches)
+            .context("Construct Create operation")?;
+        Self::op_to_transaction(operation).context("Construct create transaction")
+    }
+
+    /// <https://identity.foundation/sidetree/api/#sidetree-operations>
+    async fn submit_transaction(&self, tx: DIDMethodTransaction) -> Result<Value> {
+        let op = Self::op_from_transaction(tx)
+            .context("Convert DID method transaction to Sidetree operation")?;
+        let endpoint = self
+            .endpoint
+            .as_ref()
+            .ok_or(anyhow!("Missing Sidetree REST API endpoint"))?;
+        let url = format!("{}operations/", endpoint);
+        let client = Client::builder().build().context("Build HTTP client")?;
+        let resp = client
+            .post(url)
+            .json(&op)
+            .header("Accept", "application/json")
+            .header("User-Agent", ssi::USER_AGENT)
+            .send()
+            .await
+            .context("Send HTTP request")?;
+        if let Err(e) = resp.error_for_status_ref() {
+            let err: SidetreeError = resp
+                .json()
+                .await
+                .context("Transaction submit failed. Unable to read HTTP response JSON")?;
+            bail!("Transaction submit failed: {}: {}", e, err);
+        }
+        if resp.content_length() == Some(0) {
+            // Update operation may return empty body with 200 OK.
+            return Ok(Value::Null);
+        }
+        let bytes = resp.bytes().await.context("Unable to read HTTP response")?;
```

Comment:
Why not use `.json()`?

Reply:
`.json()` could be used. I think the current way is leftover from when I was debugging why certain results failed JSON parsing. (It was because the body was empty, now caught above).